From 21070457ab04c6cd2b721c86874abece4a298734 Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Sun, 2 Jul 2023 17:27:41 +0800
Subject: [PATCH 608/780] [Huawei] psi: add struct psi_group_ext

Offering: HULK
HULK inclusion
category: feature
bugzilla: 188780

-------------------------------

Change struct psi_group directly will causes the kabi broken.
Therefore, add a new struct psi_group_ext for new variables, which
will be added in the next patch of pressure.stat.

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 include/linux/psi_types.h |  8 ++++++++
 init/Kconfig              | 10 ++++++++++
 kernel/sched/psi.c        | 37 +++++++++++++++++++++++++++++++++++++
 3 files changed, 55 insertions(+)

diff --git a/include/linux/psi_types.h b/include/linux/psi_types.h
index 3bc764162ced..ec8ba7428c50 100644
--- a/include/linux/psi_types.h
+++ b/include/linux/psi_types.h
@@ -228,6 +228,14 @@ struct psi_group {
 	u64 polling_until;
 };
 
+#ifdef CONFIG_PSI_FINE_GRAINED
+struct psi_group_ext {
+	struct psi_group psi;
+};
+#else
+struct psi_group_ext { };
+#endif /* CONFIG_PSI_FINE_GRAINED */
+
 #else /* CONFIG_PSI */
 
 struct psi_group { };
diff --git a/init/Kconfig b/init/Kconfig
index b20908365ebe..637f58699233 100644
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -654,6 +654,16 @@ config PSI_CGROUP_V1
 
          Say N if unsure.
 
+config PSI_FINE_GRAINED
+	bool "Support fine grained psi under cgroup v1 and system"
+	default n
+	depends on PSI
+	help
+	  If set, fine grained pressure stall information tracking will
+	  be used for cgroup v1 and system, such as memory reclaim,
+	  memory compact and so on.
+	  Say N if unsure.
+
 endmenu # "CPU/Task time and stats accounting"
 
 config CPU_ISOLATION
diff --git a/kernel/sched/psi.c b/kernel/sched/psi.c
index 17679cce4029..5b6188e21a71 100644
--- a/kernel/sched/psi.c
+++ b/kernel/sched/psi.c
@@ -192,6 +192,24 @@ struct psi_group psi_system = {
 	.pcpu = &system_group_pcpu,
 };
 
+#ifdef CONFIG_PSI_FINE_GRAINED
+/* System-level fine grained pressure and stall tracking */
+struct psi_group_ext psi_stat_system = { };
+
+struct psi_group_ext *to_psi_group_ext(struct psi_group *psi)
+{
+	if (psi == &psi_system)
+		return &psi_stat_system;
+	else
+		return container_of(psi, struct psi_group_ext, psi);
+}
+#else
+static inline struct psi_group_ext *to_psi_group_ext(struct psi_group *psi)
+{
+	return NULL;
+}
+#endif
+
 static void psi_avgs_work(struct work_struct *work);
 
 static void poll_timer_fn(struct timer_list *t);
@@ -1024,16 +1042,31 @@ void psi_memstall_leave(unsigned long *flags)
 #ifdef CONFIG_CGROUPS
 int psi_cgroup_alloc(struct cgroup *cgroup)
 {
+#ifdef CONFIG_PSI_FINE_GRAINED
+	struct psi_group_ext *psi_ext;
+#endif
+
 	if (static_branch_likely(&psi_disabled))
 		return 0;
 
+#ifdef CONFIG_PSI_FINE_GRAINED
+	psi_ext = kzalloc(sizeof(struct psi_group_ext), GFP_KERNEL);
+	if (!psi_ext)
+		return -ENOMEM;
+	cgroup->psi = &psi_ext->psi;
+#else
 	cgroup->psi = kzalloc(sizeof(struct psi_group), GFP_KERNEL);
 	if (!cgroup->psi)
 		return -ENOMEM;
 
+#endif
 	cgroup->psi->pcpu = alloc_percpu(struct psi_group_cpu);
 	if (!cgroup->psi->pcpu) {
+#ifdef CONFIG_PSI_FINE_GRAINED
+		kfree(psi_ext);
+#else
 		kfree(cgroup->psi);
+#endif
 		return -ENOMEM;
 	}
 	group_init(cgroup->psi);
@@ -1049,7 +1082,11 @@ void psi_cgroup_free(struct cgroup *cgroup)
 	free_percpu(cgroup->psi->pcpu);
 	/* All triggers must be removed by now */
 	WARN_ONCE(cgroup->psi->poll_states, "psi: trigger leak\n");
+#ifdef CONFIG_PSI_FINE_GRAINED
+	kfree(to_psi_group_ext(cgroup->psi));
+#else
 	kfree(cgroup->psi);
+#endif
 }
 
 /**
-- 
2.17.1

